resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/SparkRackResolver.scala

```diff
@@ -38,3 +46,72 @@ private[yarn] class SparkRackResolver {
   }
 
 }
+
+/**
+ * Utility to resolve the rack for hosts in an efficient manner.
+ * It will cache the rack for individual hosts to avoid
+ * repeatedly performing the same expensive lookup.
+ *
+ * This will be unnecessary in hadoop releases with YARN-9332.
+ * With that, we could just directly use [[org.apache.hadoop.yarn.util.RackResolver]].
+ * In the meantime, this is a re-implementation for spark's use.
+ */
+object SparkRackResolver extends Logging {
```

Comment:
The reason why I mentioned `SparkRackResolver` was to add this functionality to the existing class, which can be easily overridden in tests where needed.

Adding a new `object` defeats that purpose. This should be added to the existing class, where you can also remove the `init()` hack and do that in the constructor.

Reply:
`init()` has a parameter `conf: Configuration` shouldn't bond to an instance of `SparkRackResolver`, so it can't move to the constructor. And move the `resolveRacks` to class instead of object will add a constructor parameter in `YarnScheduler`:
```
class YarnScheduler(sc: SparkContext, resolver: SparkRackResolver)
```

Reply:
Your init only does anything the first time it's called. I don't get why is that different from passing the configuration in the constructor, or even have the current way of just having it as part of the `resolve()` method and have it be ignored in subsequent calls.

Reply:
I will move `init()` as part of `resolver()`